sim/midas/src/main/scala/midas/models/dram/IngressUnit.scala

```diff
@@ -94,28 +106,60 @@ class IngressModule(val cfg: BaseConfig)(implicit val p: Parameters) extends Mod
                        ((awCredits.value < wCredits.value) && awCredits.inc) ||
                         awCredits.inc && wCredits.inc
 
+  val read_req_done = arQueue.io.enq.fire()
   when (!io.relaxed) {
     Seq(awCredits, wCredits) foreach { _.dec := write_req_done }
   }
 
+  val readMatcher = Module(new FIFOAddressMatcher(cfg.maxReads*2, addrWidth)).io
+  readMatcher.enq.valid := !ingressUnitStall && io.nastiInputs.hBits.ar.valid
+  readMatcher.enq.bits := io.nastiInputs.hBits.ar.bits.addr
+  readMatcher.match_address := io.nastiInputs.hBits.aw.bits.addr
 
-  val read_req_done = arQueue.io.enq.fire()
+  val writeMatcher = Module(new FIFOAddressMatcher(cfg.maxWrites*2, addrWidth)).io
+  writeMatcher.enq.valid := !ingressUnitStall && io.nastiInputs.hBits.aw.valid
+  writeMatcher.enq.bits := io.nastiInputs.hBits.aw.bits.addr
+  writeMatcher.match_address := io.nastiInputs.hBits.aw.bits.addr
 
+  arIDQueue.io.enq.valid := tFireHelper.fire(arIDQueue.io.enq.ready) && io.nastiInputs.hBits.ar.valid
+  arIDQueue.io.enq.bits := io.nastiInputs.hBits.ar.bits.id
+  awIDQueue.io.enq.valid := tFireHelper.fire(awIDQueue.io.enq.ready) && io.nastiInputs.hBits.aw.valid
+  awIDQueue.io.enq.bits := io.nastiInputs.hBits.aw.bits.id
   // FIFO that tracks the relative order of reads and writes are they are received 
   // bit 0 = Read, bit 1 = Write
   val xaction_order = Module(new DualQueue(Bool(), cfg.maxReads + cfg.maxWrites))
-  xaction_order.io.enqA.valid := read_req_done
+  val xaction_order_collisions = Module(new DualQueue(Bool(), cfg.maxReads + cfg.maxWrites))
```

Comment:
Add a comment explaining what this structure does (as above).

Reply:
Added a comment